### PR TITLE
SID: cuda temporaries

### PR DESCRIPTION
**Pull request by havogt**
Introduces SID for CUDA temporaries. Alignment and padding will be implemented during integration, when we can measure performance of different approaches. See comment in the code.

### CODE REVIEW

**Comment by havogt** [comment]
launch strgrid

File: include/gridtools/stencil_composition/backend_cuda/tmp_storage_sid.hpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        int_t ExtentJPlus,
+        uint_t NColors,
+        class Allocator,
+        class BlockSizeI = integral_constant<int_t, ComputeBlockSizeI - ExtentIMinus + ExtentIPlus>,
+        class BlockSizeJ = integral_constant<int_t, ComputeBlockSizeJ - ExtentJMinus + ExtentJPlus>,
+        class Strides = hymap::keys<dim::i, dim::j, dim::c, tmp_cuda::block_i, tmp_cuda::block_j, dim::k>::values<
+            integral_constant<int_t, 1>,
+            BlockSizeI,
+            integral_constant<int_t, BlockSizeI{} * BlockSizeJ{}>,
+            integral_constant<int_t, BlockSizeI::value * BlockSizeJ::value * NColors>,
```
Comment (havogt) [inline include/gridtools/stencil_composition/backend_cuda/tmp_storage_sid.hpp]: If I change `BlockSizeI::value` to `BlockSizeI{}` nvcc does a weird code replacement which produces an error.

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

File: unit_tests/stencil_composition/test_simple_device_memory_allocator.cu [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        TEST(simple_device_memory_allocator, test) {
+            simple_device_memory_allocator alloc;
+            auto ptr_holder = alloc.allocate<float_type>(1);
+
+            bool result = false;
+            bool *dev_result;
+            GT_CUDA_CHECK(cudaMalloc(&dev_result, sizeof(bool)));
+            GT_CUDA_CHECK(cudaMemcpy(dev_result, &result, sizeof(bool), cudaMemcpyHostToDevice));
+
+            test_allocated<<<1, 1>>>(ptr_holder, dev_result);
```
Comment (anstaf) [inline unit_tests/stencil_composition/test_simple_device_memory_allocator.cu]: you can use `on_device::exec` here

File: unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cu [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include "test_tmp_storage_sid_cuda.cpp"
+
+#include "../../test_helper.hpp"
+#include <gridtools/stencil_composition/backend_cuda/simple_device_memory_allocator.hpp>
+
+namespace gridtools {
+    namespace {
+        using tmp_cuda_storage_sid_float = tmp_cuda_storage_sid<float_type, simple_device_memory_allocator>;
+        TEST_F(tmp_cuda_storage_sid_float, maker_with_device_allocator) {
+            auto testee = get_tmp_storage();
```
Comment (anstaf) [inline unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cu]: This is confusing.

You use test fixture just to get the type of the returned sid.
Can you do it explicitly :
```
  using testee_t = decltype(make_tmp_storage_cuda<int>(
                    tmp_cuda::blocksize<1, 1>{},
                    extent<>{},
                    0,
                    0,
                    0,
                    std::declval<simple_device_memory_allocator&>()));
```
?
And move `static_asserts` out of the test function? 

If you include `*.cpp` file, please include it the last line of the file.

File: unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        constexpr int_t extent_i_minus = -1;
+        constexpr int_t extent_i_plus = 2;
+        constexpr int_t extent_j_minus = -3;
+        constexpr int_t extent_j_plus = 4;
+
+        constexpr int_t blocksize_i = 32;
+        constexpr int_t blocksize_j = 8;
+
+        template <typename T, typename Allocator> // test fixture works with default constructible allocators
+        class tmp_cuda_storage_sid : public ::testing::Test {
```
Comment (anstaf) [inline unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cpp]: I don't see the point of writing test fixture here.
You just distribute input parameters for your the only one testing scenario over three different places -- global staff; members of the fixture; local variables in the test function.

File: unit_tests/stencil_composition/icosahedral_grids/test_tmp_storage_sid_cuda.cu [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+            using tmp_cuda_t = decltype(testee);
+
+            static_assert(sid::is_sid<tmp_cuda_t>::value, "is_sid()");
+
+            ASSERT_TYPE_EQ<GT_META_CALL(sid::ptr_type, tmp_cuda_t), data_t *>();
+            ASSERT_TYPE_EQ<GT_META_CALL(sid::ptr_diff_type, tmp_cuda_t), int_t>();
+        }
+    } // namespace
+} // namespace gridtools
```
Comment (anstaf) [inline unit_tests/stencil_composition/icosahedral_grids/test_tmp_storage_sid_cuda.cu]: The same comments as for strgrid stuff.

I would prefer to be consistent here -- if you have a single `*.hpp` file with `icgrid/strgrid` branches driven by defines, do the same in testing.

File: unit_tests/stencil_composition/icosahedral_grids/test_tmp_storage_sid_cuda.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                                    sid::shift(ptr, host::at_key<dim::j>(strides), j);
+                                    sid::shift(ptr, host::at_key<dim::c>(strides), c);
+                                    sid::shift(ptr, host::at_key<tmp_cuda::block_i>(strides), bi);
+                                    sid::shift(ptr, host::at_key<tmp_cuda::block_j>(strides), bj);
+                                    sid::shift(ptr, host::at_key<dim::k>(strides), k);
+                                    ASSERT_EQ((index_info{i, j, c, bi, bj, k}), *ptr);
+                                }
+        }
+    } // namespace
+} // namespace gridtools
```
Comment (anstaf) [inline unit_tests/stencil_composition/icosahedral_grids/test_tmp_storage_sid_cuda.cpp]: The same comments as for strgrid stuff.

File: unit_tests/cuda_test_helper.hpp [hunk 7]
```diff
@@ -15,6 +15,10 @@
 
 #pragma once
 
+#ifndef __CUDACC__
+#error This is CUDA only header
+#endif
+
```
Comment (anstaf) [inline unit_tests/cuda_test_helper.hpp]: restore new line at the file end please

Comment (havogt) [reply unit_tests/cuda_test_helper.hpp]: I think it is there, the deleted line was an additional new line.

File: unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cpp (hunk 4 above)
Comment (havogt) [reply unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cpp]: Yes, the test fixture is an artifact of the previous test structure...

File: unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cu (hunk 3 above)
Comment (havogt) [reply unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cu]: Do you think I should include the .cpp test in this case?

Comment (anstaf) [reply unit_tests/stencil_composition/structured_grids/test_tmp_storage_sid_cuda.cu]: may be. it verifies nvcc compilation

File: unit_tests/tools/multiplet.hpp [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <gridtools/common/defs.hpp>
+#include <gridtools/common/host_device.hpp>
+
+/**
+   @brief Small value type to use in tests where we want to check the
+   values in a fields, for instance to check if layouts works, on in
+   communication tests
+*/
+
```
[the remark this answers is not in the document]
Comment (havogt) [reply unit_tests/tools/multiplet.hpp]: I'll keep it for now.

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch icgrid

**Pull request merged**